---
title: 'COOL NAME'
author: "Amy M. Van Cise^1\\*^, Mollie Ball^1^, OTHERS" # FIX ORDER

output: 
    html_document:
      fig_caption: yes

bibliography: "`r rbbt::bbt_write_bib('bibliography.json', overwrite = TRUE)`"
# csl: royal-society-open-science.csl

knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile, encoding = encoding, output_dir = "docs",
  output_file = "index.html") })
---

1. School of Aquatic and Fishery Sciences, University of Washington, Seattle, WA
2. 
3. 

<br>

^\*^Corresponding author email: avancise\@gmail.com 

<br>

Running page head: *Arctic Predator Metabarcoding*

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE)

library(tidyverse)
library(PNWColors)
library(viridis)
library(pander)
library(english)
library(magrittr)
library(rbbt)
library(patchwork)
library(png)

load(".Rdata")
load(".Rdata")
load(".Rdata")
load("")
load("")
load("")
```

## ABSTRACT {-#ABSTRACT}


<br>

Key Words: arctic predators, metabarcoding, microbiome, diet composition, marine mammals

<br>

## INTRODUCTION {-#INTRODUCTION}


## METHODS {-#METHODS}

### Sample collection
We receieved (*xx*) stomach and fecal samples from ringed seals (*n=XX*), bearded seals (*n=XX*), and beluga whales (*n=XX*) from the Alaskan Department of Fish and Game (ADFG). (*LOCATIONS*), (*SAMPLES WERE FROM STRANGINGS AND INDIGENOUS HARVEST??*) <-- INCLUDE MORE INFORMATION HERE.

### DNA Extraction and Sequencing
- 
- How much sample volume did we use?
- QIAmp Mini Stool Kit <-- did we modify?
"*Whole genomic DNA was extracted from swabbed, wet scooped, and lyophilized samples using the QIAmp Fast DNA Stool Mini Kit (Qiagen). . We quantified the amount of DNA extracted from each sample, then amplified the 16S mitochondrial region using published primers and protocols (Van Cise et al. 2024). We determined PCR success through visual examination using gel electrophoresis, and re-amplified unsuccessful samples under a variety of parameters to increase sample success rate. Those parameters included: 1:10 dilution, 1:2 dilution, DNA cleaned using a Zymo Clean and Concetrate kit, and 2x concetration of cleaned DNA. After all iterations of PCR1 were run, we collected those samples with successfully amplified product to continue through the library preparation protocol." *

*Next we cleaned the amplified PCR product using the AMPure bead clean protocol and indexed each sample using Unique Dual Indexes (UDI, Illumina, Inc). ... We then performed a second AMPure bead clean to remove all unused reagents, and pooled all samples to contain an equimolar mix of prey DNA from each sample.*

*Once the amplified product was pooled, we performed a final gel cleaning step using the MinElute PCR purificaton kit (Qiagen, Inc) to remove any non-target DNA fragments, then diluted and denatured the DNA according to the MiSeq sequencing protocol and loaded the pooled DNA and 20% PhiX control onto an in-house MiSeq sequencer.*"
 - Sophie




### Data analysis

All data analyses and plots were generated using the R coding language in R studio [@RCoreTeam2016]. 

## RESULTS {-#RESULTS}


<br>
```{r farm.location, fig.height = 6, fig.width = 8, fig.align = 'center', fig.cap = "Figure 1. Location of the two farm sites used in this study, near the Romero and Media Luna reefs off the southwestern coast of Puerto Rico in the Gulf of Mexico."}

map

```

```{r data_summary, fig.height = 6, fig.width = 8, fig.align = 'center', fig.cap = "Figure 2. Top: Recording effort in each month and year at both farm sites. Middle: Normalized number of farm interactions detected in each month and year throughout the study period. Bottom: Duration of farm interactions over the study period. The large figure highlights the duration of most farm interactions, while the inset shows outlier interactions."}

effort + eventsMonthly + eventDur_study + 
  plot_layout(ncol = 1, guides = "collect") &
  theme(legend.position = "right")

```

<br>

```{r detect_prob, fig.align = 'center', fig.cap = "Figure 3. Predicted binomial probability of the occurence of a farm interaction at each farm site. The left plot shows the proportion of click-positive minutes observed at each farm location. The right shows predicted probability of detection based on GAM model output. The top panel shows the effect of diel proximity to sunrise or sunset on interaction probability at each farm site, while the bottom panel shows the effect of Julian Day on interaction probability at each farm site."}

p1 <- (dielDetect_farm_plot + seasonalDetect_farm_plot) + plot_layout(ncol = 1, axes = "collect_y", guides = "collect")

eventNumloc + p1
```

<br>


<br>

## DISCUSSION {-#DISCUSSION}
 
## CONCLUSION {-}

## ACKNOWLEDGEMENTS
We are grateful to [FIELD CREW NAMES HERE] for their data collected onsite in .
    
## REFERENCES {-#REFERENCES}
<div id="refs"></div>

<br>

<br>

## SUPPLEMENTAL MATERIALS
